pkg/bitmap/bitmap.go

```diff
@@ -0,0 +1,237 @@
+// Copyright 2021 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package bitmap provides the implementation of bitmap.
+package bitmap
+
+import (
+	"math"
+	"math/bits"
+)
+
+// Bitmap struct contains numOnes which represent number of ones in the
+// bitmap and a slice bitBlock which represent bits. The type of bitBlock
+// is uint64 which means each number in bitBlock can contain information
+// for 64 bits.
+//
+// +stateify savable
+type Bitmap struct {
+	numOnes  uint32
+	bitBlock []uint64
+}
+
+// NewBitmap create a new empty Bitmap.
+func NewBitmap(size uint32) *Bitmap {
+	b := new(Bitmap)
+	bSize := size / 64
+	if size % 64 != 0 {
+		bSize = bSize + 1
+	}
+	b.bitBlock = make([]uint64, bSize, bSize)
+	return b
+}
+
+// IsEmpty verifiys if the Bitmap is empty.
+func (b *Bitmap) IsEmpty() bool {
+	return b.numOnes == 0
+}
+
+// Minimum return the smallest value in the Bitmap.
+func (b *Bitmap) Minimum() uint32 {
+	for i := 0; i < len(b.bitBlock); i++ {
+		if w := b.bitBlock[i]; w != 0 {
+			r := bits.TrailingZeros64(w)
+			return uint32(r + i*64)
+		}
+	}
+	return math.MaxInt32
+}
+
+// Maximum return the largest value in the Bitmap.
+func (b *Bitmap) Maximum() uint32 {
+	for i := len(b.bitBlock); i > 0; i-- {
+		if w := b.bitBlock[i-1]; w != 0 {
+			r := bits.LeadingZeros64(w)
+			return uint32((i-1)*64 + 63 - r)
+		}
+	}
+	return uint32(0)
+}
+
+// Add add i to the Bitmap.
+func (b *Bitmap) Add(i uint32) {
+	blockNum, mask := i/64, uint64(1)<<(i%64)
+	// if blockNum is out of range, extend b.bitBlock
+	if x, y := int(blockNum), len(b.bitBlock); x >= y {
+		b.bitBlock = append(b.bitBlock, make([]uint64, x-y+1)...)
+	}
+	oldBlock := b.bitBlock[blockNum]
+	newBlock := oldBlock | mask
+	if oldBlock != newBlock {
+		b.bitBlock[blockNum] = newBlock
+		b.numOnes++
+	}
+}
+
+// Remove i from the Bitmap.
+func (b *Bitmap) Remove(i uint32) {
+	blockNum, mask := i/64, uint64(1)<<(i%64)
+	oldBlock := b.bitBlock[blockNum]
+	newBlock := oldBlock &^ mask
+	if oldBlock != newBlock {
+		b.bitBlock[blockNum] = newBlock
+		b.numOnes--
+	}
+}
+
+// Clone the Bitmap.
+func (b *Bitmap) Clone() *Bitmap {
+	ptr := Bitmap{b.numOnes, make([]uint64, len(b.bitBlock))}
+	copy(ptr.bitBlock, b.bitBlock[:])
+	return &ptr
+}
+
+// ReplaceWith replace the content of b with that of newBitmap.
```

Comment:
Is this called anywhere? If not, it can be deleted.

Reply:
This function was supposed to be used in `loadFdBitmap` which is state loader for fdBitmap in fd_table.go.
I will discuss this with you in the comment in `fd_table.go`